src/pybind/mgr/volumes/fs/operations/versions/subvolume_v2.py

```diff
@@ -0,0 +1,418 @@
+import os
+import stat
+import uuid
+import errno
+import logging
+from datetime import datetime
+
+import cephfs
+
+from .metadata_manager import MetadataManager
+from .subvolume_base import SubvolumeBase
+from .subvolume_base import SubvolumeTypes
+from .subvolume_base import SubvolumeFeatures
+from ..op_sm import SubvolumeOpSm
+from ..op_sm import SubvolumeStates
+from ..template import SubvolumeTemplate
+from ..snapshot_util import mksnap, rmsnap
+from ...exception import IndexException, OpSmException, VolumeException, MetadataMgrException
+from ...fs_util import listdir
+from ..template import SubvolumeOpType
+
+from ..clone_index import open_clone_index, create_clone_index
+
+log = logging.getLogger(__name__)
+
+class SubvolumeV2(SubvolumeBase, SubvolumeTemplate):
+    VERSION = 2
+
+    @staticmethod
+    def version():
+        return SubvolumeV2.VERSION
+
+    @property
+    def path(self):
+        try:
+            # no need to stat the path -- open() does that
+            return self.metadata_mgr.get_global_option(MetadataManager.GLOBAL_META_KEY_PATH).encode('utf-8')
+        except MetadataMgrException as me:
+            raise VolumeException(-errno.EINVAL, "error fetching subvolume metadata")
+
+    @property
+    def features(self):
+        return [SubvolumeFeatures.FEATURE_SNAPSHOT_CLONE.value,
+                SubvolumeFeatures.FEATURE_SNAPSHOT_AUTOPROTECT.value,
+                SubvolumeFeatures.FEATURE_SNAPSHOT_RETENTION.value]
+
+    def _set_incarnation_metadata(self, subvolume_type, qpath, initial_state):
+        self.metadata_mgr.update_global_section(MetadataManager.GLOBAL_META_KEY_TYPE, subvolume_type.value)
+        self.metadata_mgr.update_global_section(MetadataManager.GLOBAL_META_KEY_PATH, qpath)
+        self.metadata_mgr.update_global_section(MetadataManager.GLOBAL_META_KEY_STATE, initial_state.value)
+
+    def create(self, size, isolate_nspace, pool, mode, uid, gid):
+        subvolume_type = SubvolumeTypes.TYPE_NORMAL
+        try:
+            initial_state = SubvolumeOpSm.get_init_state(subvolume_type)
+        except OpSmException as oe:
+            raise VolumeException(-errno.EINVAL, "subvolume creation failed: internal error")
+
+        subvol_path = os.path.join(self.base_path, str(uuid.uuid4()).encode('utf-8'))
+        try:
+            # TODO: decide if subvolume rather than the incarnation should hold the size, nspace, pool attributes
+            # and incarnation should hold mode, uid, gid
+            self.fs.mkdirs(subvol_path, mode)
+            self.set_attrs(subvol_path, size, isolate_nspace, pool, uid, gid)
+
+            # persist subvolume metadata
+            qpath = subvol_path.decode('utf-8')
+            try:
+                self.metadata_mgr.refresh()
```

Comment:
Would it help if we reach here after an open() design wise? So the metadata is already readout.

Reply:
The issue that I see is, we would need to expose state to [`create_subvolume`](https://github.com/ceph/ceph/blob/7fb21a25e1cdcce27af0f0f9b040aa1537fb42fc/src/pybind/mgr/volumes/fs/volume.py#L160-L168), such that it can decide to call create or work with the opened subvolume on open success. Similarly for `clone_subvolume_snapshot`.

It looks better to abstract away this detail for users of Subvolume*.

If we decide to add "--force", then `create_subvolume` can decide based on open and state if this should be allowed, but again this can also be abstracted away to the Subvolume implementation, to fail the Subvolume.create if "--force" is not used.

Thoughts?

Reply:
> The issue that I see is, we would need to expose state to [`create_subvolume`](https://github.com/ceph/ceph/blob/7fb21a25e1cdcce27af0f0f9b040aa1537fb42fc/src/pybind/mgr/volumes/fs/volume.py#L160-L168), such that it can decide to call create or work with the opened subvolume on open success. Similarly for `clone_subvolume_snapshot`.

state need not be passed back to create_subvol() -- just that create() would be valid on a valid subvolume object.

Would that help with the resize case where attr changes can be done? That could possibly reduce abstract methods such as resize, etc..

Reply:
open() returning a success for both ENOENT and EEXIST would help, with create picking up the idempotent requirements. IOW, create() handles updating attrs on an existing subvolume or creating one afresh.

Although I would not want to add it to this PR, considering its current size and scope.